test/parallel-benchmark/mzcompose.py

```diff
@@ -0,0 +1,1190 @@
+        start_time: float,
+        conns: queue.Queue,
+        measurements: collections.defaultdict[str, list[Measurement]],
+    ):
+        self._run(conns)
+        duration = time.time() - start_time
+        measurements[str(self)].append(Measurement(duration, start_time))
+
+    def _run(self, conns: queue.Queue):
+        raise NotImplementedError
+
+
+class TdAction(Action):
+    def __init__(self, td: str, c: Composition):
+        self.td = dedent(td)
+        self.c = c
+
+    def _run(self, conns: queue.Queue):
+        self.c.testdrive(self.td, quiet=True)
+
+    def __str__(self) -> str:
+        return "testdrive"
+
+
+class StandaloneQuery(Action):
+    def __init__(
+        self,
+        query: str,
+        conn_info: PgConnInfo,
+        strict_serializable: bool = True,
+    ):
+        self.query = query
+        self.conn_info = conn_info
+        self.strict_serializable = strict_serializable
+
+    def _run(self, conns: queue.Queue):
+        conn = self.conn_info.connect()
+        with conn.cursor() as cur:
+            if not self.strict_serializable:
+                cur.execute("SET TRANSACTION_ISOLATION TO 'SERIALIZABLE'")
+            cur.execute(self.query)
+        conn.close()
+
+    def __str__(self) -> str:
+        return f"{self.query} (standalone)"
+
+
+class ReuseConnQuery(Action):
+    def __init__(
+        self,
+        query: str,
+        conn_info: PgConnInfo,
+        strict_serializable: bool = True,
+    ):
+        self.query = query
+        self.conn = conn_info.connect()
+        self.conn.autocommit = True
+        self.cur = self.conn.cursor()
+        if not strict_serializable:
+            self.cur.execute("SET TRANSACTION_ISOLATION TO 'SERIALIZABLE'")
+
+    def _run(self, conns: queue.Queue):
+        self.cur.execute(self.query)
+
+    def __str__(self) -> str:
+        return f"{self.query} (reuse connection)"
+
+
+class PooledQuery(Action):
+    def __init__(self, query: str, conn_info: PgConnInfo):
+        self.query = query
+        self.conn_info = conn_info
+
+    def _run(self, conns: queue.Queue):
+        conn = conns.get()
+        with conn.cursor() as cur:
+            cur.execute(self.query)
+        conns.task_done()
+        conns.put(conn)
+
+    def __str__(self) -> str:
+        return f"{self.query} (pooled)"
+
+
+def sleep_until(timestamp: float) -> None:
+    time_to_sleep = timestamp - time.time()
+    if time_to_sleep > 0:
+        time.sleep(time_to_sleep)
+
+
+class Distribution:
+    def generate(self, duration: int) -> Iterator[float]:
+        raise NotImplementedError
+
+
+class Periodic(Distribution):
+    """Run the action in each thread in one second, spread apart by the 1/per_second"""
+
+    def __init__(self, per_second: int):
+        self.per_second = per_second
+
+    def generate(self, duration: int) -> Iterator[float]:
+        next_time = time.time()
+        for i in range(duration * self.per_second):
+            yield time.time()
```

Comment:
I think this should yield `next_time` instead of the current time, then we're actually modelling the external system correctly. The key insight is that we're never in control about when requests arrive, so we need to use `next_time`, which is a representative of the request produced by the external system.